### PR TITLE
Jest: Added Generic for "toMatchObject" function

### PR DESCRIPTION
**Pull request by favna**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: **None. It is a purely internal change not based on any changes to the lib.**
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

File: types/jest/index.d.ts [hunk 1]
```diff
@@ -757,8 +758,25 @@ declare namespace jest {
...
+         *     amenities: string[];
+         *     area: number;
+         *     wallColor: string;
+         *   }
+         * };
+         *
+         * expect(desiredHouse).toMatchObject<House>(...standardHouse, kitchen: {area: 20}) // wherein standardHouse is some base object of type House
          */
-        toMatchObject(expected: {} | any[]): R;
+        toMatchObject<O extends {}>(expected: O | any[]): R;
```
Comment (wsmd) [inline types/jest/index.d.ts]: `toMatchObject` can also work with arrays:

> You can also pass an array of objects, in which case the method will return true only if each object in the received array matches (in the toMatchObject sense described above) the corresponding object in the expected array.

So we could probably support arrays in the generic as well. For example, currently the following will pass without any issues.

```ts
.toMatchObject<[{ foo: string }, { bar: number }]>([{ foo: true }, { bar: () => {} }]); // no errors!
```

What do you think about something like this? This should add type safety for tuples and array types via the generic if desired.

```suggestion
        toMatchObject<E extends {} | any[]>(expected: E): R;
```

File: types/jest/index.d.ts [hunk 2]
```diff
@@ -757,8 +758,25 @@ declare namespace jest {
         toMatch(expected: string | RegExp): R;
         /**
          * Used to check that a JavaScript object matches a subset of the properties of an object
+         *
+         * Optionally, you can provide an object to use as Generic type for the expected value.
+         * This ensures that the matching object matches the structure of the provided object-like type.
```
Comment (wsmd) [inline types/jest/index.d.ts]: I don't see any comments similar to this in here.

I believe the documentation throughout the file is intended to document Jest's API as stated in its docs, not on how to use the types. The type information, such as generic arguments, can still be inferred from the editor sometimes:

Screen Shot 2019-04-27 at 11 28 34 AM

**Review by wsmd — CHANGES_REQUESTED** [review]
Thanks for the PR @Favna! This will be a helpful addition.

I left a couple of notes. Let me know what you think!

File: types/jest/index.d.ts (hunk 2 above)
Comment (Favna) [reply types/jest/index.d.ts]: I based the including of an example on `.toHaveNthReturnedWith()` typing just above `toMatchObject()`. As for inferring from IntelliSense, the examples are in fact also included in the IntelliSense and I personally thing it's nice to have some extra documentation, especially for TypeScript structures that are considered "advanced" (being that Generics are under "advanced types" on the docs) so it helps TS starters.

GIF recording of the IntelliSense I see, with the example:

19-57-46_27-04-2019_%pn


TL;DR: I think it doesn't give any negative effects and only (if minor), positive ones. I see no reason to remove it.

File: types/jest/index.d.ts (hunk 1 above)
Comment (Favna) [reply types/jest/index.d.ts]: That's a very good suggestion. I've made your changes as well as added a corresponding test in dbfa03d852c22dbe723f559a8029ce389278570a

File: types/jest/index.d.ts (hunk 2 above)
Comment (wsmd) [reply types/jest/index.d.ts]: The comment added stood out a little bit to me from the rest of the documentation comments since it isn't describing Jest's functionality, but given what you said above I don't think it hurts to keep it either. Thanks for the clarification! 😄

**Review by wsmd — APPROVED** [review]
LGTM! 👍

File: types/jest/index.d.ts [hunk 3]
```diff
@@ -757,8 +758,25 @@ declare namespace jest {
...
+         *   bath: boolean;
+         *   bedrooms: number;
+         *   kitchen: {
+         *     amenities: string[];
+         *     area: number;
+         *     wallColor: string;
+         *   }
+         * };
+         *
+         * expect(desiredHouse).toMatchObject<House>(...standardHouse, kitchen: {area: 20}) // wherein standardHouse is some base object of type House
```
Comment (jwbay) [inline types/jest/index.d.ts]: Is the intent to spread into an object here?
```suggestion
         * expect(desiredHouse).toMatchObject<House>({ ...standardHouse, kitchen: {area: 20} }) // wherein standardHouse is some base object of type House
```

Comment (Favna) [reply types/jest/index.d.ts]: This was indeed meant to be an object spread. I indeed forgot a `{`. Looks like the PR was merged already though so this'll have to be covered in a next one. That said... It's not *really* important I think so I think I'll just wait for the next PR to come up and the bot will ping me now as well so I can request the change then from whoever is editing.
